File: demos/test-ts/src/algorithms/codingInverview/buildTree.ts
import { TreeNode } from '../TreeNode';

// 递归的方法
// 先序遍历:  [根, [左], [右]]
// 中序遍历:  [[left], root, [right]]
export function buildTree(
  preorder: number[],
  inorder: number[]
): TreeNode<number> {
  const n = preorder.length;
  const indexMap = new Map<number, number>();

  for (let i = 0; i < n; i++) {
    indexMap.set(inorder[i], i);
  }
  function myBuildTree(
    preorder: number[],
    inorder: number[],
    preLeft: number,
    preRight: number,
    inLeft: number,
    inRight: number
  ): TreeNode<number> {
    if (preLeft > preRight) return null;

    // pre root
    let preRootIdx = preLeft;

    const inRootIdx = indexMap.get(preorder[preRootIdx]);

    const root: TreeNode<number> = new TreeNode(preorder[preRootIdx]);

    const leftSize = inRootIdx - inLeft;
    root.left = myBuildTree(
      preorder,
      inorder,
      preLeft + 1,
      preLeft + leftSize,
      inLeft,
      inRootIdx - 1
    );
    root.right = myBuildTree(
      preorder,
      inorder,
      preLeft + 1 + leftSize,
      preRight,
      inRootIdx + 1,
      inRight
    );
    return root;
  }
  return myBuildTree(preorder, inorder, 0, n - 1, 0, n - 1);
}
// 递归法
// Rust 可以使用这种方式
// 使用一个 Data 结构体存储  p 和 d. 或者将 p ,d 当作参数进行递归调用
export function buildTree4(preorder: number[], inorder: number[]) {
  const p: IterableIterator<number> = preorder[Symbol.iterator]();
  const d: Map<number, number> = new Map(inorder.map((v, k) => [v, k]));
  const helper = (i: number, j: number): TreeNode<number> | null => {
    if (i < j) {
      const val = p.next().value,
        idx = <number>d.get(val);
      return new TreeNode(val, helper(i, idx), helper(idx + 1, j));
    }
    return null;
  };
  return helper(0, inorder.length);
}

// 迭代法
export function buildTree2(
  preorder: number[],
  inorder: number[]
): TreeNode<number> {
  if (!preorder || preorder.length === 0) {
    return null;
  }
  const root = new TreeNode(preorder[0]);

  const stack: TreeNode<number>[] = [];
  stack.push(root);
  // 中序索引
  let inorderIndex = 0;
  // 遍历先序
  for (let i = 1; i < preorder.length; i++) {
    const val = preorder[i];
    let node = stack[stack.length - 1];

    if (node.val !== inorder[inorderIndex]) {
      node.left = new TreeNode(val);
      stack.push(node.left);
    } else {
      while (
        !!stack.length &&
        stack[stack.length - 1].val === inorder[inorderIndex]
      ) {
        node = stack.pop();
        inorderIndex++;
      }
      node.right = new TreeNode(val);
      stack.push(node.right);
    }
  }
  return root;
}

// 速度偏慢. Rust 可以使用这种递归的方法
export function buildTree3(preorder: number[], inorder: number[]) {
  function helper(preorder: number[], inorder: number[]): TreeNode<number> {
    if (preorder && preorder.length) {
      const first = preorder[0];
      const pivotIdx = inorder.findIndex((x) => x === first);
      const node = new TreeNode(first);
      node.left = helper(
        preorder.slice(1, 1 + pivotIdx),
        inorder.slice(0, pivotIdx)
      );
      node.right = helper(
        preorder.slice(1 + pivotIdx),
        inorder.slice(1 + pivotIdx)
      );
      return node;
    } else {
      return null;
    }
  }
  return helper(preorder, inorder);
}

File: demos/test-rust/coding_interview/src/build_tree.rs
// use std::cell::RefCell;
// use std::rc::Rc;
use std::collections::HashMap;
use std::{cell::RefCell, rc::Rc};

use crate::base_st::TreeNode;

pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
    if preorder.len() == 0 {
        return None;
    }
    let n = preorder.len();
    let mut mp: HashMap<i32, i32> = HashMap::new();
    for i in 0..n {
        // 注意是中序的 HashMap
        mp.insert(inorder[i], i as i32);
    }
    fn recur(
        preorder: &Vec<i32>,
        inorder: &Vec<i32>,
        pre_left: i32,
        pre_right: i32,
        in_left: i32,
        in_right: i32,
        mp: &HashMap<i32, i32>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if pre_left > pre_right {
            return None;
        }
        let pre_root_idx = pre_left;
        let in_root_idx = *mp.get(&preorder[pre_root_idx as usize]).unwrap();

        let left_size = in_root_idx - in_left;
        let root_val = preorder[pre_root_idx as usize];
        let root = Rc::new(RefCell::new(TreeNode::new(root_val)));

        root.borrow_mut().left = recur(
            preorder,
            inorder,
            pre_left + 1,
            pre_left + left_size,
            in_left,
            in_root_idx - 1,
            mp,
        );
        root.borrow_mut().right = recur(
            preorder,
            inorder,
            pre_left + left_size + 1,
            pre_right,
            in_root_idx + 1,
            in_right,
            mp,
        );
        Some(root)
    }

    recur(
        &preorder,
        &inorder,
        0,
        (n - 1) as i32,
        0,
        (n - 1) as i32,
        &mp,
    )
}

// 这个是别人的写法
pub fn build_tree2(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
    helper(&preorder, &inorder)
    // 在 Solution 里面可以这么写
    // Self::helper(&preorder, &inorder)
}

fn helper(preorder: &[i32], inorder: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
    if let Some(root) = preorder.first() {
        let pivot_idx = inorder
            .iter()
            .enumerate()
            .find(|(_, v)| v == &root)
            .unwrap()
            .0;
        return Some(Rc::new(RefCell::new(TreeNode {
            val: *root,
            left: helper(&preorder[1..(1 + pivot_idx)], &inorder[0..pivot_idx]),
            right: helper(&preorder[(1 + pivot_idx)..], &inorder[(pivot_idx + 1)..]),
        })));
    } else {
        return None;
    }
}
